[1/2] discovery+lnwire: add support for DNS host name in NodeAnnouncement msg

mohamedawnallah opened the pull request:
## Change Description

Towards #6337.
Towards #9126.
Next #10159.

## Steps to Test
Steps for reviewers to follow to test the change.

## Pull Request Checklist
### Testing
- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [ ] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [ ] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [ ] Any new logging statements use an appropriate subsystem and logging level.
- [ ] Any new lncli commands have appropriate tags in the comments for the rpc in the proto file.
- [ ] [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

alexbosworth commented on graph/db/addr.go:
```diff
@@ -26,6 +27,9 @@ const (
 	// v3OnionAddr denotes a version 3 Tor (prop224) onion service address.
 	v3OnionAddr addressType = 3
+
+	// dnsHostnameAddr denonates a DNS hostname address.
```
```suggestion
	// dnsHostnameAddr denotes a DNS hostname address.
```

saubyk commented on sample-lnd.conf:
```diff
@@ -214,6 +214,11 @@
+
+; Specify a DNS hostname for the node's external address. If no port is
+; provided, the default (9735) is used.
+; external-dns-hostname=
```
can add an example here like "lightning.mynode.org" or something

mohamedawnallah replied:
Thanks that would be helpful. I've used the same domain example `my-node-domain.com` used in `externalhosts` flag for consistency.

ellemouton commented on lnrpc/lightning.proto:
```diff
@@ -3437,6 +3437,9 @@ message LightningNode {
     map<uint64, bytes> custom_records = 7;
+
+    // The DNS hostname address of the node, if available.
+    NodeAddress dns_hostname_address = 8;
```
why a new field for this if we already have an `addresses` field above?

ellemouton commented on cmd/commands/peersrpc_active.go:
```diff
@@ -69,6 +69,10 @@ var updateNodeAnnouncementCommand = cli.Command{
 			Name:  "color",
```
a suggestion for PR structure: it's a better pattern to do:
1) add new logic internal to the server
2) expose new logic of server via RPC
3) start reading/using new logic from client

a nice way to think about it is: we always want to be able to revert back to any commit and have things make sense. But with this current order, if we reverted to this commit, we would have all these fields that the server is exposing and the client is writing to but it wont actually do anything on the server side which doesnt make sense

ellemouton commented on config.go:
```diff
@@ -325,26 +325,28 @@ type Config struct {
 	// loadConfig function. We need to expose the 'raw' strings so the
```
this commit is doing way too much. 
- start with just `lnwire` (ie, "schema first" rule) and make sure to add the necessary tests around decoding/encoding there. Also, maybe i missed this but somewhere we need to check that our peers are not sending us more than 1 DNS address in their announcements (see bolt 7). 
- then slowly expand out from there, commit by commit to start using the new type. 

each commit should be one unit of change that the reviewer needs to consider.

lightninglabs-deploy commented:
@mohamedawnallah, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@mohamedawnallah, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@mohamedawnallah, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@mohamedawnallah, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@mohamedawnallah, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@mohamedawnallah, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@mohamedawnallah, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@mohamedawnallah, remember to re-request review from reviewers when ready

mohamedawnallah commented:
This PR is not yet ready for review. It looks like more code paths need to be test-covered. I also understand the git commit structure is not still followed per latest feedback as I was trying to make CI happy. I will take a look at this PR soon.

mohamedawnallah commented:
Thank you, @ellemouton, for your feedback on this PR.

I have restructured the commits and added unit tests for encoding/decoding the DNS hostname address, as well as for parsing it. Apologies for the significant delay between receiving your review and addressing it—I’ll do my best to minimize such gaps in the future, in line with the `New Development Process`.

Any follow-up feedback would be very much appreciated!

PS:
It looks like the failed CI tests not related to the changes in this PR also It doesn't look like we need to add itests for this feature change, no?

ellemouton commented on lnwire/dns_hostname_addr.go:
```diff
@@ -0,0 +1,26 @@
+)
+
+// DNSHostnameAddress is a custom implementation of the net.Addr interface.
+type DNSHostnameAddress struct {
```
This PR doesnt seem to add a case for this type in `lnwire.ReadElement`. This means that if we get any node announcements that include type 5 (ie, DNS), we wont parse them as such and will continue to include the bytes in the `OpaqueAddrs` type which we dont currently resolve at all as this holds address bytes we assume we cannot parse. 

We should definitely have a test case to ensure that are properly taking a raw node_announcement byte stream and correctly parsing it to and from the `lnwire.NodeAnnouncement` type.

mohamedawnallah replied:
> This PR doesnt seem to add a case for this type in `lnwire.ReadElement`. This means that if we get any node announcements that include type 5 (ie, DNS), we wont parse them as such and will continue to include the bytes in the `OpaqueAddrs` type which we dont currently resolve at all as this holds address bytes we assume we cannot parse.

That’s right. I’m curious—when a new address needs to be supported, it looks like it has to be added twice:

- once in `graph/db/addr.go`
- and again in `lnwire/lnwire`.

It also seems likely that it needs to be tested twice. Am I missing anything?

ellemouton reviewed (changes requested):
Thanks for the PR :) 

I think its a good idea to test/ensure the following: 
1) make sure the node can handle providing its own DNS addr
2) make sure the node can handle receiving a DNS addr from other nodes

mohamedawnallah commented on sample-lnd.conf:
```diff
@@ -214,6 +214,14 @@
+; Default:
+;   external-dns-hostname=
+; Example:
+;   external-dns-hostname=my-node-domain.com
```
I was able to reproduce this error as well. I'll add itests to my toolbox wherever it makes sense. 👍

mohamedawnallah commented on server_test.go:
```diff
@@ -80,3 +129,228 @@ func TestShouldPeerBootstrap(t *testing.T) {
+	}{
+		{
+			//nolint:ll
+			name:    "OnionAddress_WithExplicitPort_ReturnsOnionAddr",
```
I think the choice of test case naming is somewhat subjective, but I personally find this style more readable. By using the `<Category>_<Condition>_<ExpectedResult>` pattern, I can quickly see what is being tested, under what condition, and what the expected outcome is—all from the test name itself.

For example, with the current naming:

```
$ go test -v -run "TestParseAddress"
--- PASS: TestParseAddress (0.00s)
    --- PASS: TestParseAddress/OnionAddress_WithExplicitPort_ReturnsOnionAddr (0.00s)
    --- PASS: TestParseAddress/OnionAddress_WithoutPort_ReturnsOnionAddrWithDefaultPort (0.00s)
    --- PASS: TestParseAddress/LoopbackAddress_WithExplicitPort_ReturnsTCPAddr (0.00s)
    --- PASS: TestParseAddress/LoopbackAddress_WithoutPort_ReturnsTCPAddrWithDefaultPort (0.00s)
    --- PASS: TestParseAddress/IPv4Address_WithExplicitPort_ReturnsTCPAddr (0.00s)
    --- PASS: TestParseAddress/IPv4Address_WithoutPort_ReturnsTCPAddrWithDefaultPort (0.00s)
    --- PASS: TestParseAddress/IPv6Address_WithExplicitPort_ReturnsTCPAddr (0.00s)
    --- PASS: TestParseAddress/IPv6Address_WithoutPort_ReturnsTCPAddrWithDefaultPort (0.00s)
    --- PASS: TestParseAddress/DNSHostname_WithExplicitPort_ReturnsDNSHostnameAddress (0.00s)
    --- PASS: TestParseAddress/DNSHostname_WithoutPort_ReturnsDNSHostnameAddressWithDefaultPort (0.00s)
    --- PASS: TestParseAddress/Hostname_DNSLookupFailure_ReturnsError (0.00s)
    --- PASS: TestParseAddress/DNSHostname_InvalidFormat_ReturnsError (0.00s)
    --- PASS: TestParseAddress/Address_NonNumericPort_ReturnsPortParsingError (0.00s)
    --- PASS: TestParseAddress/Address_EmptyString_ReturnsInvalidAddressError (0.00s)
    --- PASS: TestParseAddress/IPAddress_TCPResolutionFailure_ReturnsError (0.00s)
```

This makes each test clear and descriptive, but I'm open to adjusting the style if a simpler naming convention is preferred!

mohamedawnallah commented on server_test.go:
```diff
@@ -80,3 +129,228 @@ func TestShouldPeerBootstrap(t *testing.T) {
+		netCfg  tor.Net
+		want    net.Addr
+		wantErr bool
+		errMsg  string
```
I think keeping them separate provides more clarity, even though the state of `errMsg` depends on `wantErr`.

If they’re combined into something like `errMsg` or `expectedErrMsg` (with `""` as the default), it might create ambiguity for someone reading the test cases. For example:

- Does an empty `errMsg` mean that the test case encountered an error, but the error message wasn’t captured?
- Or does it mean there was no error at all?

Keeping `wantErr` and `errMsg` distinct helps avoid this kind of confusion.

What do you think?

mohamedawnallah commented on graph/db/addr.go:
```diff
@@ -31,6 +31,9 @@ const (
 	// opaqueAddrs denotes an address (or a set of addresses) that LND was
 	// not able to parse since LND is not yet aware of the address type.
 	opaqueAddrs addressType = 4
+
+	// dnsHostnameAddr denotes a DNS hostname address.
+	dnsHostnameAddr addressType = 5
```
That makes sense. I'm new to the practice of adding migration scripts. My understanding is that we need migrations whenever we make changes that aren't backward-compatible. In our case, without migration, there would be parsing errors since DNS addresses currently stored as type 4 would incorrectly end up in the opaque address bucket instead of being properly recognized as DNS addresses (type 5). Is my understanding correct?

mohamedawnallah replied:
So would this migration script be located in something like `channeldb/migration34`? And from a user perspective, they would typically learn about this migration through the release notes?
```bash
package migration

import (
    "fmt"
    "github.com/lightningnetwork/lnd/kvdb"
)

// MigrateDNSAddressTypes converts node announcements with DNS addresses
// from type 4 to type 5 to ensure proper decoding.
func MigrateDNSAddressTypes(tx kvdb.RwTx) error {
    // Get the node announcement bucket
    nodeBucket := tx.ReadWriteBucket(nodeBucket)
    if nodeBucket == nil {
        return fmt.Errorf("node bucket not found")
    }

    // Iterate through all node announcements
    return nodeBucket.ForEach(func(k, v []byte) error {
        if v == nil {
            return nil
        }

        // Decode the node announcement
        node, err := deserializeNode(v)
        if err != nil {
            return err
        }

        needsUpdate := false
        
        // Check addresses for any DNS addresses incorrectly stored as type 4
        for i, addr := range node.Addresses {
            // Check if this is a DNS address incorrectly stored as type 4
            if addr.Type == 4 && isDNSAddress(addr) {
                // Update to type 5 (DNS hostname address type)
                node.Addresses[i].Type = 5
                needsUpdate = true
            }
        }

        // If we found addresses that need updating, save the changes
        if needsUpdate {
            // Serialize the updated node
            updatedBytes, err := serializeNode(node)
            if err != nil {
                return err
            }

            // Write back to the database
            return nodeBucket.Put(k, updatedBytes)
        }

        return nil
    })
}
```

mohamedawnallah replied:
Also what should we do if a node in the channel graph sends us multiple DNS addresses that were previously stored as opaque addresses? How should we handle this scenario in the migration?

mohamedawnallah commented on lnwire/dns_hostname_addr_test.go:
```diff
@@ -0,0 +1,171 @@
+// non-ASCII characters, control characters, and various valid hostname formats
+// and port combinations.
+func TestDNSHostnameAddressValidate(t *testing.T) {
+	tests := []struct {
```
Will add `t.Parallel()` later along with the migration script 👍

mohamedawnallah commented on graph/db/kv_store.go:
```diff
@@ -4087,11 +4087,24 @@ func deserializeLightningNode(r io.Reader) (models.LightningNode, error) {
+					"Announcement message. See Bolt 7")
+
+				return models.LightningNode{}, err
+			}
```
> note that this is also only checking this at read time. Should be checked before write time cause we dont want to persist the message in the first place if it violates spec rules


In addition to checking read/write at lnwire layer. Should we also check for duplicate DNS addresses here? we could make it take a list of `[]net.Addr` with func name `SerializeAddresses` so we could have flag state if DNS repeated more than once. What do you think?
https://github.com/lightningnetwork/lnd/blob/3707b1fb7032d98dc2a8e881d3ae83e61fbc35e3/graph/db/addr.go#L257-L270

mohamedawnallah reviewed (commented):
Thanks @ellemouton for the detailed feedback. I've left some follow-up comments and questions for clarification 🙏

GustavoStingelin commented on lnwire/dns_hostname_addr.go:
```diff
@@ -0,0 +1,96 @@
+// Lightning Network's BOLT 07 specification requirements. It ensures the
+// hostname is not empty, doesn't exceed 255 characters, contains only ASCII
+// characters, and that the port number is within the valid TCP range (1-65535).
+func (d *DNSHostnameAddress) Validate() error {
```
What if, instead of using a Validate function, we introduce a Parse() function that returns either a valid DNSHostnameAddress or an error? This would help prevent the creation of invalid states, allowing us to always trust that the struct contains valid data.

For reference, this is the same pattern used in the Go standard library, e.g., [netip.ParseAddr](https://pkg.go.dev/net/netip#ParseAddr). It encourages a more data-oriented and error-defensive design.

More specifically, when we expose the Hostname field publicly without enforcing construction through Parse or MustParse, we’re implicitly allowing invalid values (e.g., a hostname with millions of characters). Only later do we catch errors via Validate, which can lead to inconsistencies and bugs. This touches on a crucial aspect of data semantics: making invalid states unrepresentable by construction.

mohamedawnallah replied:
> What if, instead of using a Validate function, we introduce a Parse() function that returns either a valid DNSHostnameAddress or an error? This would help prevent the creation of invalid states, allowing us to always trust that the struct contains valid data.

I like this idea I used a similar way `NewDNSHostnameAddress` let me know what do you think :)

``` go
func NewDNSHostnameAddress(host string, port int) (*DNSHostnameAddress, error) {
	dnsAddr := &DNSHostnameAddress{
		Hostname: host,
		Port:     port,
	}

	if err := dnsAddr.validate(); err != nil {
		return nil, err
	}

	return dnsAddr, nil
}
func (d *DNSHostnameAddress) validate() error {
...
}
```

> More specifically, when we expose the Hostname field publicly without enforcing construction through Parse or MustParse, we’re implicitly allowing invalid values (e.g., a hostname with millions of characters). Only later do we catch errors via Validate, which can lead to inconsistencies and bugs. This touches on a crucial aspect of data semantics: making invalid states unrepresentable by construction.

I understand the concern about exposing Hostname publicly, but enforcing strict construction through `Parse` or `MustParse` isn’t strictly necessary. Many standard Go packages, like `net.TCPAddr` and `net.IP`, expose public fields and allow manual construction of potentially invalid states. These remain reliable because validation happens during critical I/O operations, such as `Dial` or Listen.

In this case, `NewDNSHostnameAddress` ensures validity at creation, and while `Hostname` can be modified later, this flexibility is often useful. Invalid states are only an issue if they’re used without validation, which can be addressed with proper documentation and validation during I/O operations. This also the similar pattern used in handling onion address:

https://github.com/lightningnetwork/lnd/blob/c52a6ddeb8e596c6f4fcd61382d2a0a0f6ee9abc/tor/onionaddr.go#L41-L51

GustavoStingelin replied:
I'm fine with this for now.

mohamedawnallah commented:
Drafting it for now to handle a regression for considering invalid ip addresses to be valid. Also use the suggested function design from @GustavoStingelin using `Parse` instead of  OOP way of using `Validate`.

mohamedawnallah commented:
> Drafting it for now to handle a regression for considering invalid ip addresses to be valid.

The main point is that an IP address like `500.0.0.0` is considered **syntactically valid as a DNS hostname** according to RFC specifications. 

For example, using a library such as [miekg/dns](https://github.com/miekg/dns) with code like:

```go
dns.IsDomainName("500.0.0.0")
```

returns `true`.

This can feel odd in certain cases, such as when a user updates their announcement with that address and sees it added. For example:

```bash
lncli peers updatenodeannouncement --address_add 500.0.0.0
```

This might seem strange to the user.

So, it’s unclear whether this should be considered a **feature** or a **bug** 😅.  

However, based on both DNS standard rules and Bolt-07 rules, this is **objectively valid** since the hostname consists of ASCII characters within the valid length limit.

mohamedawnallah commented:
> Drafting it for now to handle a regression for considering invalid ip addresses to be valid.

The main point is that an IP address like `500.0.0.0` is considered **syntactically valid as a DNS hostname** according to RFC specifications. 

For example, using a library such as [miekg/dns](https://github.com/miekg/dns) with code like:

```go
dns.IsDomainName("500.0.0.0")
```

returns `true`.

This can feel odd in certain cases, such as when a user updates their announcement with that address and sees it added. For example:

```bash
lncli peers updatenodeannouncement --address_add 500.0.0.0
```

This might seem strange to the user.

So, it’s unclear whether this should be considered a **feature** or a **bug** 😅.  

However, based on both DNS standard rules and Bolt-07 rules, this is **objectively valid** since the hostname consists of ASCII characters within the valid length limit.

mohamedawnallah commented:
Hey @ellemouton,

I’ve addressed the feedback and added a few follow-up questions for clarification. Would you be able to take another look when you have some time?

Thanks!

ellemouton reviewed (commented):
Very nice work @mohamedawnallah !

Left some comments. 

I did some testing of this (after first rebasing on https://github.com/lightningnetwork/lnd/pull/9856) and can confirm it works nicely!

From the testing I can see that: if a node has a node announcement stored that has a DNS addr inside and then starts up, we still cant see it and still see `unknown network for unrecognized address type`. However, this then gets fixed as soon as we get an updated node address from the node in question since then we go overwrite all its old addresses and properly use the new type. 

So I think we could actually get away with doing the migration in a separate PR to this one since I think it is not critical to migrate this data since things will be consistent eventually anways if the node announcement is received again. All it means is that if we were to ship with only this change sans the migration then if a node has a DNS addr persisted of another node, it wont be able to properly read/see that until it receives a new node announcement to override that. 

I think it would be good to get a second opinion on this point :)

yyforyongyu reviewed (changes requested):
Love how much test-oriented this PR is! My main question is about the usage of the DNS - from a user's perspective, do we allow more than one DNS to be saved on disk? And how can a user update its DNS record?

GustavoStingelin commented on server.go:
```diff
@@ -528,16 +528,33 @@ func parseAddr(address string, netCfg tor.Net) (net.Addr, error) {
-	return netCfg.ResolveTCPAddr("tcp", hostPort)
+	// For loopback or IP addresses: Use ResolveTCPAddr to properly
+	// resolve these through Tor or other proxies, preventing IP leakage.
+	if lncfg.IsLoopback(host) || isIP(host) || lnwire.IsIPv4Like(host) {
```
If `500.0.0.0` is a DSN that will fail the resolution in the future, why do we need to know if it's "like an IPv4"? I think with this function, we're introducing some unclear and ambiguous logic, bc I need to stop, go down to the function, and learn what "like an IP" means.

GustavoStingelin replied:
I understand the point about preventing 500.0.0.0 at an early stage, but if we proceed to resolve DNS and it fails, we can skip this logic and rely on the DNS resolution itself as the validation mechanism. WDYT?

GustavoStingelin replied:
Perhaps we could always call netCfg.ResolveTCPAddr after confirming the address is not an Onion address. This function handles DNS resolution when the provided address is not a valid IP and returns an error if the resolution fails. With this approach, parseAddr will return an error when appropriate, preventing the server from starting with an invalid address. What do you think of this approach?

GustavoStingelin replied:
That's a good point. My idea is to use the resolver only to check if the DNS is valid, without resolving it to actual IP addresses. The algorithm would look like this:
```
if isOnion {
    return OnionAddr
}
if isIPv4OrIPv6 {
    return TCPAddr
}

// It's a DNS. We create a new DNS struct, but first, we validate if the DNS is reachable
if err := netCfg.ResolveTCPAddr(...); err != nil {
    return err // DNS resolution failed, so we return an error
}

// Other validations for local IPs could go here, but adding them might overload the function with too many responsibilities

// DNS is valid
return DNSAddr
```

but on the other hand, we have the function lncfg.ParseAddressString that does similar logic, working on https://github.com/lightningnetwork/lnd/issues/8801 I feel that perhaps we should refactor both functions to improve consistency

lightninglabs-deploy commented:
@ellemouton: review reminder
@mohamedawnallah, remember to re-request review from reviewers when ready

ellemouton commented on lnwire/dns_addr.go:
```diff
@@ -0,0 +1,146 @@
+// Lightning Network's BOLT 07 specification requirements. It ensures the
+// hostname is not empty, doesn't exceed 255 characters, contains only ASCII
+// characters, and that the port number is within the valid TCP range (1-65535).
+func (d *DNSAddr) validate() error {
```
i still dont think we need all this validation on this layer. I think it's ok to do all the validation for our _own_ users hostname string at config load time or on the update-nod-ann rpc functionality. But the spec only says "The hostname must be ASCII and not longer than 258 chars" so imo that's the only validation to do here. 
wdyt @yyforyongyu ?

ellemouton commented on server.go:
```diff
@@ -529,16 +529,54 @@ func parseAddr(address string, netCfg tor.Net) (net.Addr, error) {
+	}
+
+	return addr, nil
+}
```
ah ok, i didnt see that this was what was passed to the peer rpc server

ellemouton commented on lnwire/lnwire.go:
```diff
@@ -884,6 +885,49 @@ func ReadElement(r io.Reader, element interface{}) error {
+						"multiple DNS addresses. See " +
+						"Bolt 07")
+				}
+				dnsAddrIncluded = true
```
> the restriction one the announcement should happen in the gossip, not here?

not sure i agree: the `lnwire` package defines valid protocol level messages and it is invalid for a node_announcement message to contain more than 1 DNS addr & so i think it makes sense to fail at decode time if this is violated.

ellemouton replied:
>  I don't think this approach works since after a restart the user can save another DNS record again.

Not sure i follow this - can you expand?

ellemouton commented on lnwire/lnwire.go:
```diff
@@ -881,6 +885,49 @@ func ReadElement(r io.Reader, element interface{}) error {
+				}
+
+				length := 1 + uint16(hostnameLen) + portLen
+				addrBytesRead += length
```
1. 
> Also DNS doesn't have mapping here since it has variable hostname

note that in the code snippet i gave, `AddrLen` was adapted to take in a param to account for the variable length. 

2. 
>  perform DNS validation in a centralized pla

The spec only says we need to check for valid ascii. The other checks are not spec related. 

3. 
> s the difference between suggested one and existing 

There are various other simplifications. Note that there is great merit in keeping the code style the same as the other cases in this switch as it makes the code more readable.

ellemouton commented on graph/db/addr.go:
```diff
@@ -150,6 +150,43 @@ func encodeOpaqueAddrs(w io.Writer, addr *lnwire.OpaqueAddrs) error {
+// representation. It writes the address type, hostname length, hostname, and
+// port (in big-endian order) to the writer. The function validates that the
+// DNS address. Returns an error if any part of the serialization fails.
+func encodeDNSAddr(w io.Writer, addr *lnwire.DNSAddr) error {
```
as discussed offline & in previous review comment: it is going to be much simpler to not touch the persistence layer. Rather just dynamically extract DNS addrs at read time from the existing opaque type & similarly at write time, add any DNS types to the opaque stream.

ellemouton replied:
because also, note that as it currently stands: any existing DNS addrs that we have stored already will currently be under the opaque addr type - so you have to account for those at read-time anyways.

ellemouton commented on graph/db/addr.go:
```diff
@@ -31,6 +31,9 @@ const (
 	// not able to parse since LND is not yet aware of the address type.
 	opaqueAddrs addressType = 4
+
+	// dnsAddr denotes a DNS hostname address.
+	dnsAddr addressType = 5
```
note: this is unused in this commit

lightninglabs-deploy commented:
@mohamedawnallah, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@mohamedawnallah, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@mohamedawnallah, remember to re-request review from reviewers when ready

GustavoStingelin commented on lnwire/dns_addr.go:
```diff
@@ -0,0 +1,68 @@
+
+// Validate validates that the DNS hostname is not empty and contains only ASCII
+// characters and of max length 255 characters according to BOLT specifications.
+func (d *DNSAddr) Validate() error {
```
IMO, we should add validation to ensure the port is non-zero and fits within the 2-byte range. According to BOLT 7, type 1, type 2, and type 5 address descriptors MUST NOT have a port equal to 0, and ports larger than 2 bytes are invalid.

GustavoStingelin commented on lnwire/dns_addr.go:
```diff
@@ -0,0 +1,68 @@
+func (d *DNSAddr) Validate() error {
+	// Check if hostname is empty.
+	if d.Hostname == "" {
+		return errors.New("hostname cannot be empty")
```
minor improvement suggestion: if we define these errors at the package level, it becomes easier to match them in test cases using `errors.Is` or `errors.As`

GustavoStingelin commented on server.go:
```diff
@@ -547,16 +547,59 @@ func parseAddr(address string, netCfg tor.Net) (net.Addr, error) {
+		return nil, err
+	}
+
+	// Check if that DNS address resolve to any TCP addresses.
+	if _, err = netCfg.ResolveTCPAddr("tcp", addr.String()); err != nil {
+		return nil, err
+	}
```
👍🏼

GustavoStingelin commented on config.go:
```diff
@@ -1614,6 +1617,17 @@ func ValidateConfig(cfg Config, interceptor signal.Interceptor, fileParser,
 
+		// Parse the external DNS address if provided.
+		if cfg.RawExternalDNSAddress != "" {
+			addr, err := parseDNSAddr(
+				cfg.RawExternalDNSAddress, cfg.net,
+			)
+			if err != nil {
+				return nil, err
+			}
+			cfg.ExternalDNSAddress = addr
```
nit: missing new line

GustavoStingelin reviewed (changes requested):
Just a few comments, but I'm ok with the current approach! 🚀

ellemouton commented:
@mohamedawnallah  - please see the failing itests as they are related to this PR

ellemouton commented:
@mohamedawnallah , you need to run with the same build flags that the tests that are failing are using. Ie, there are failing with the sql graph store. The issue will be related to the other comments i left on the PR with regards to persistence

ellemouton commented:
ok just spoke to @saubyk  and he had a good idea: perhaps we are overcomplicating this & we can potentially get away with _not_ unraveling anything we have already persisted on disk (ie, Opaque Addrs remain Opaque addrs even if they contain DNS addrs). I think this might be an ok solution given that nodes will occasionally send out new announcements & so things should _eventually_ be correct. 

Im going to start a convo offline just to first see what others think. In the mean time, defs take a look at the draft code i linked before to get an understanding of what that path looks like.

yyforyongyu reviewed (approved):
LGTM🚢 Left some nits - will merge once CI passed
